src/PostgREST/Query.hs

```diff
@@ -50,50 +52,52 @@ import PostgREST.Plan                    (ActionPlan (..),
                                           InfoPlan (..),
                                           InspectPlan (..))
 import PostgREST.Plan.MutatePlan         (MutatePlan (..))
-import PostgREST.Plan.ReadPlan           (ReadPlanTree)
 import PostgREST.Query.SqlFragment       (escapeIdentList, fromQi,
                                           intercalateSnippet,
                                           setConfigWithConstantName,
                                           setConfigWithConstantNameJSON,
                                           setConfigWithDynamicName)
-import PostgREST.Query.Statements        (ResultSet (..))
+import PostgREST.Query.Statements        (PreparedStatement (..),
+                                          ResultSet (..))
 import PostgREST.SchemaCache             (SchemaCache (..))
 import PostgREST.SchemaCache.Identifiers (QualifiedIdentifier (..))
-import PostgREST.SchemaCache.Routine     (MediaHandler, Routine (..),
-                                          RoutineMap)
+import PostgREST.SchemaCache.Routine     (Routine (..), RoutineMap)
 import PostgREST.SchemaCache.Table       (TablesMap)
 
 import Protolude hiding (Handler)
 
 type DbHandler = ExceptT Error SQL.Transaction
 
+data Query
+  = DbQuery {
+      dqIsoLevel    :: SQL.IsolationLevel,
+      dqTxMode      :: SQL.Mode,
+      dqDbHandler   :: DbHandler QueryResult,
+      dqTransaction :: SQL.IsolationLevel -> SQL.Mode -> SQL.Transaction (Either Error QueryResult) -> SQL.Session (Either Error QueryResult),
+      dqSQL         :: ByteString
+    }
+  | NoDbQuery QueryResult
+
```

Comment:
:eyes: These new types are looking good! If possible, consider adding them on a `refactor` commit so this PR is easier to review.